heron/metricsmgr/src/java/org/apache/heron/metricsmgr/sink/metricscache/MetricsCacheClient.java

```diff
@@ -93,12 +93,20 @@ private void addMetricsCacheClientTasksOnWakeUp() {
     Runnable task = new Runnable() {
       @Override
       public void run() {
-        while (!publishMetricsCommunicator.isEmpty()) {
-          TopologyMaster.PublishMetrics publishMetrics = publishMetricsCommunicator.poll();
+        TopologyMaster.PublishMetrics publishMetrics;
+        synchronized (publishMetricsCommunicator) {
+          publishMetrics = publishMetricsCommunicator.poll();
+        }
```

Comment:
hmm, if we change the while loop condition, maybe we could reduce the code duplicate here?

Reply:
like while (synchronized (publishMetricsCommunicator) { publishMetrics = .... } ) {}?

As a non-java programmer, I don't know if it is ok. :D 

But this is a valid point. Let me see if I can improve it.

Reply:
👍